### PR TITLE
Fix save/load/resume from checkpoint for DeepSpeed Plugin

### LINKED ISSUE #7282: Model restore fails from stored checkpoint when using Deepspeed
## 🐛 Bug
Trying to restore a checkpoint to resume training but it fails with the below exceptions

```
RuntimeError: Error(s) in loading state_dict for BoringModel:
        size mismatch for layer.weight: copying a param with shape torch.Size([2, 32]) from checkpoint, the shape in current model is torch.Size([1]).
        size mismatch for layer.bias: copying a param with shape torch.Size([2]) from checkpoint, the shape in current model is torch.Size([1]).
    self.lightning_module.load_state_dict(ckpt['state_dict'])
  File "/home/ca5b7a03-2d901b-2d45e5-2d969e-2df8ccc075972b/.local/lib/python3.7/site-packages/torch/nn/modules/module.py", line 1052, in load_state_dict
    self.__class__.__name__, "\n\t".join(error_msgs)))
RuntimeError: Error(s) in loading state_dict for BoringModel:
        size mismatch for layer.weight: copying a param with shape torch.Size([2, 32]) from checkpoint, the shape in current model is torch.Size([1]).
        size mismatch for layer.bias: copying a param with shape torch.Size([2]) from checkpoint, the shape in current model is torch.Size([1]).
    self.__class__.__name__, "\n\t".join(error_msgs)))
RuntimeError: Error(s) in loading state_dict for BoringModel:
        size mismatch for layer.weight: copying a param with shape torch.Size([2, 32]) from checkpoint, the shape in current model is torch.Size([1]).
        size mismatch for layer.bias: copying a param with shape torch.Size([2]) from checkpoint, the shape in current model is torch.Size([1]).
    self.__class__.__name__, "\n\t".join(error_msgs)))
RuntimeError: Error(s) in loading state_dict for BoringModel:
        size mismatch for layer.weight: copying a param with shape torch.Size([2, 32]) from checkpoint, the shape in current model is torch.Size([1]).
        size mismatch for layer.bias: copying a param with shape torch.Size([2]) from checkpoint, the shape in current model is torch.Size([1]).
```

### To Reproduce
Run the following mode with commented out restore argument, then run it again with uncommenting the restore and you will see the exception.

```python
import os
import torch
from torch.utils.data import Dataset, DataLoader
from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.callbacks import ModelCheckpoint
from pytorch_lightning.plugins import DeepSpeedPlugin
from deepspeed.ops.adam import FusedAdam


class RandomDataset(Dataset):

    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):

    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return loss

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)
        return loss

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)
        return loss

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    test_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    checkpoint_callback = ModelCheckpoint(
        dirpath='tests/',
        filename='{epoch:02d}',
    )
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        gpus=-1,
        limit_train_batches=1,
        limit_val_batches=1,
        num_sanity_val_steps=0,
        precision=16,
        accelerator='ddp',
        max_epochs=2,
        plugins=[DeepSpeedPlugin(cpu_offload=False, stage=3)],
        weights_summary=None,
        callbacks=[checkpoint_callback],
        #resume_from_checkpoint='tests/epoch=01.ckpt',
    )
    trainer.fit(model, train_dataloader=train_data, val_dataloaders=val_data)
    trainer.test(model, test_dataloaders=test_data)


if __name__ == '__main__':
    run()

```

### Expected behavior

Training resume successfully from stored checkpoint.

### Environment

Tried with lightning version: `1.2.10, 1.3.0.rc1 and master`
pytorch: `1.7.1`
OS: `Ubuntu 18.04`

@SeanNaren As discussed on slack  ^^

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Closes #7282 https://github.com/PyTorchLightning/pytorch-lightning/issues/8344 https://github.com/PyTorchLightning/pytorch-lightning/issues/8092

Re-write of #7297

This PR will fix DeepSpeed saving/loading and resuming.

TLDR: DeepSpeed ZeRO Stage 3 was broken when saving/loading. This is now fixed. If the user wants to resume from checkpoint they need to set `DeepSpeedPlugin(save_full_weights=False)` when using ZeRO Stage 3. Documentation will be updated once PR is merged!

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [x] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 1]
```diff
@@ -645,55 +644,97 @@ def save_checkpoint(self, checkpoint: Dict, filepath: str) -> None:
...
-        return client_state
 
     def load_model_state_dict(self, checkpoint: Mapping[str, Any]) -> None:
         # override to do nothing, deepspeed engine already loaded the weights in `load_checkpoint_file()`
-        pass
+        if self.save_full_weights and self.zero_stage_3:
+            self.model_to_device()
+            self._restore_zero_state(checkpoint)
+
+    def _restore_zero_state(self, ckpt: Mapping[str, Any]) -> None:
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/deepspeed.py:684]: Love it !

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -876,9 +866,25 @@ def _run(self, model: 'pl.LightningModule') -> Optional[Union[_EVALUATE_OUTPUT,
         # plugin will setup fitting (e.g. ddp will launch child processes)
         self._pre_dispatch()
 
+        # restore modules after setup
+        self.checkpoint_connector.resume_start()
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:870]: @awaelchli I believe you had a reason for having them before setup right ?

Comment (SeanNaren) [reply pytorch_lightning/trainer/trainer.py:870]: Yep you're right @tchaton this logic will need to be moved back, as this is necessary for the user to load weights before modules are wrapped via Fully Sharded `configure_sharded_model`.

Going to be a tough one to resolve without moving this as deepspeed is required to be initialized to load the model, but going to see what I can do

**Comment by SeanNaren** [comment]
I've updated the logic as follows:

1. User trains with DeepSpeed Plugin
2. User finishes training and checkpoints are saved
3. User tries to resume from checkpoint, and it works fine from pointing to the checkpoint directory

User then asks `how do I get a single checkpoint file?`.

Documentation points to this gist: https://gist.github.com/SeanNaren/e0535977df8e165247f7ff9ae445be1b which suggests the user to run:

```
python convert_lightning_deepspeed_checkpoint.py checkpoint_directory/ output.ckpt
```

Then the user can load the model as normal:

```python
MyModel.load_from_checkpoint('output.ckpt')
```

To load for inference when using DeepSpeed, the user must pass load_full_weights:

```python
Trainer(plugins=DeepSpeedPlugin(load_full_weights=True).predict(model, ckpt_path='output.ckpt', datamodule=dm)
```

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 3]
```diff
@@ -657,70 +664,113 @@ def _create_default_config(
...
-    def _filepath_to_dir(self, filepath: str) -> str:
-        return os.path.dirname(filepath)
-
     @property
     def deepspeed_engine(self):
         return self.model
 
+    @property
+    def _multi_device(self) -> bool:
+        return self.num_processes > 1 or self.num_nodes > 1
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/deepspeed.py:673]: Reminder to myself to check if this is even correct and add a test

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 4]
```diff
@@ -250,7 +253,7 @@ def __init__(
 
             synchronize_checkpoint_boundary: Insert :func:`torch.cuda.synchronize` at each checkpoint boundary.
 
-            save_full_weights: Gathers weights across all processes before saving to disk
+            load_full_weights: Gathers weights across all processes before saving to disk
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/deepspeed.py:256]: The description is incorrect here, also we need some way to test this (we could simulate this by training without DeepSpeed, and then trying to run predict WITH the deepspeed plugin)

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 5]
```diff
@@ -657,70 +664,113 @@ def _create_default_config(
...
-        else:
-            super().save_checkpoint(checkpoint, filepath)
-
-    def load_checkpoint_file(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
-        if self.save_full_weights or self.world_size == 1:
+        if self.zero_stage_3 and self._multi_device:
+            warning_cache.warn(
+                "When saving the DeepSpeed Stage 3 checkpoint, "
+                "each worker will save a shard of the checkpoint within a directory."
+                "If a single file is required after training, see <TODO> for instructions."
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/deepspeed.py:686]: Make sure to update documentation in this PR for reference + add a link to the appropriate section.

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 6]
```diff
@@ -657,70 +664,113 @@ def _create_default_config(
...
+
+    def load_checkpoint_file(self, checkpoint_path: Union[str, Path]) -> Optional[Dict[str, Any]]:
+        if self.load_full_weights and self.zero_stage_3:
             # Broadcast to ensure we load from the rank 0 checkpoint
             # This doesn't have to be the case when using deepspeed sharded checkpointing
             checkpoint_path = self.broadcast(checkpoint_path)
             return super().load_checkpoint_file(checkpoint_path)
 
-        # Rely on deepspeed to load the checkpoint and necessary information
+        # Rely on deepspeed completely to load the checkpoint and necessary information
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/deepspeed.py:701]:
```suggestion
        # Rely on deepspeed to load the checkpoint and necessary information
```

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 7]
```diff
@@ -657,70 +664,113 @@ def _create_default_config(
...
+                    load(child, prefix + name + ".")
+
+        load(self.lightning_module, prefix="")
 
     def load_optimizer_state_dict(self, checkpoint: Mapping[str, Any]) -> None:
         # override to do nothing, deepspeed engine already loaded the states in `load_checkpoint_file()`
-        pass
+        if self.load_full_weights and self.zero_stage_3 and self.lightning_module.trainer.state.fn == TrainerFn.FITTING:
+            rank_zero_warn(
+                "A single checkpoint file was saved using ZeRO Stage 3. This means optimizer states and "
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/deepspeed.py:769]: Message needs to be updated.

File: pytorch_lightning/trainer/trainer.py [hunk 8]
```diff
@@ -543,8 +543,6 @@ def fit(
             model, train_dataloaders=train_dataloaders, val_dataloaders=val_dataloaders, datamodule=datamodule
         )
 
-        self.checkpoint_connector.resume_start()
```
Comment (SeanNaren) [inline pytorch_lightning/trainer/trainer.py:546]: Need a separate PR for this.

File: pytorch_lightning/callbacks/model_checkpoint.py [hunk 9]
```diff
@@ -525,7 +525,7 @@ def save_function(self, value: Optional[Callable]) -> None:
 
     def _del_model(self, trainer: "pl.Trainer", filepath: str) -> None:
         if trainer.should_rank_save_checkpoint and self._fs.exists(filepath):
-            self._fs.rm(filepath)
+            self._fs.rm(filepath, recursive=True)
```
Comment (SeanNaren) [inline pytorch_lightning/callbacks/model_checkpoint.py:528]: Need to add a test for model checkpoint when using deepspeed. Since we're saving a model as a directory in DeepSpeed, needs to be recursive

File: CHANGELOG.md [hunk 10]
```diff
@@ -168,6 +168,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added support for providing callables to the Lightning CLI instead of types ([#8400](https://github.com/PyTorchLightning/pytorch-lightning/pull/8400))
 
 
+- Load ckpt path when model provided in validate/test/predict ([#8352](https://github.com/PyTorchLightning/pytorch-lightning/pull/8352)))
+
+
```
Comment (carmocca) [inline CHANGELOG.md:173]: Unrelated change?

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 11]
```diff
@@ -703,20 +701,77 @@ def load_checkpoint_file(self, checkpoint_path: Union[str, Path]) -> Dict[str, A
...
-            self.deepspeed_engine.optimizer._partition_all_parameters()
-
         _, client_state = self.deepspeed_engine.load_checkpoint(
-            save_dir, load_optimizer_states=is_fitting, load_lr_scheduler_states=is_fitting
+            checkpoint_path, load_optimizer_states=is_fitting, load_lr_scheduler_states=is_fitting
         )
+        if client_state is None:
+            raise MisconfigurationException(
+                "DeepSpeed was unable to load the checkpoint. Ensure you passed in a DeepSpeed compatible checkpoint "
+                "or a single checkpoint file with Trainer(plugins=DeepSpeedPlugin(load_full_weights=True))."
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:710]:
```suggestion
                "or a single checkpoint file with `Trainer(plugins=DeepSpeedPlugin(load_full_weights=True))`."
```

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 12]
```diff
@@ -238,11 +238,9 @@ def setup_optimizers_in_pre_dispatch(self) -> bool:
...
     def restore_checkpoint_after_pre_dispatch(self) -> bool:
         """
-        Override to delay restoring from checkpoint till after pre-dispatch.
+        Override to delay restoring from checkpoint till after predispatch.
         This is useful when the plugin requires all the setup hooks to run before loading checkpoint.
-
-        Returns:
-            If true, restore checkpoint after pre_dispatch.
+        Returns: If true, restore checkpoint after pre_dispatch.
         """
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:244]: Undo this?

The previous format looks better IMO

File: pytorch_lightning/plugins/training_type/training_type_plugin.py [hunk 13]
```diff
@@ -231,14 +231,16 @@ def setup_optimizers_in_pre_dispatch(self) -> bool:
...
+        Returns:
+            If True, delay setup optimizers till pre_dispatch, else call within setup.
         """
         return False
 
     @property
     def restore_checkpoint_after_pre_dispatch(self) -> bool:
         """
-        Override to delay restoring from checkpoint till after pre-dispatch.
+        Override to delay restoring from checkpoint till after predispatch.
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/training_type_plugin.py:243]:
```suggestion
        Override to delay restoring from checkpoint till after pre-dispatch.
```

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 14]
```diff
@@ -657,43 +663,36 @@ def _create_default_config(
...
+        return self.num_processes > 1 or self.num_nodes > 1
+
     def save_checkpoint(self, checkpoint: Dict, filepath: str) -> None:
         """Save model/training states as a checkpoint file through state-dump and file-write.
 
         Args:
             checkpoint: The checkpoint state dictionary
             filepath: write-target file's path
         """
-        if self.world_size > 1 and self.zero_stage_3:
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:674]: Why do we not want the world size checks anymore?

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 15]
```diff
@@ -703,20 +702,77 @@ def load_checkpoint_file(self, checkpoint_path: Union[str, Path]) -> Dict[str, A
...
+            self.model_to_device()
+            self._restore_zero_state(checkpoint)
+
+    def _restore_zero_state(self, ckpt: Mapping[str, Any]) -> None:
+        """
+        Overrides the normal load_state_dict behaviour in PyTorch to ensure
+        we gather parameters that may be sharded across processes before loading
+        the state dictionary when using ZeRO stage 3.
+        This is then automatically synced across processes.
+        Args:
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:738]: The docs might not render properly if there isn't a space between these (afaik)

```suggestion
        This is then automatically synced across processes.

        Args:
```

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 16]
```diff
@@ -703,20 +702,77 @@ def load_checkpoint_file(self, checkpoint_path: Union[str, Path]) -> Dict[str, A
...
+            ckpt: The ckpt file.
+        """
+
+        def load(module: torch.nn.Module, prefix=""):
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/deepspeed.py:742]: Interesting

why does deepspeed not provide this utility?

File: tests/plugins/test_deepspeed_plugin.py [hunk 17]
```diff
@@ -419,6 +417,48 @@ def test_deepspeed_fp32_works(tmpdir):
...
+        trainer.test(model, ckpt_path=checkpoint_path)
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        plugins=[DeepSpeedPlugin(stage=3, load_full_weights=True)],
+        gpus=1,
+        fast_dev_run=True,
+        precision=16,
+    )
+    trainer.test(model, ckpt_path=checkpoint_path)
```
Comment (carmocca) [inline tests/plugins/test_deepspeed_plugin.py:459]: This could use an assertion

either by accessing a property, checking caplog or mocking a function call.

File: tests/plugins/test_deepspeed_plugin.py [hunk 18]
```diff
@@ -576,16 +608,117 @@ def test_deepspeed_multigpu_stage_3_checkpointing(tmpdir):
...
+        "scheduler states can not be restored. If you'd like to restore these states, you must "
+        "provide a path to the originally saved DeepSpeed checkpoint.",
+    ):
+        trainer.fit(model, datamodule=dm)
+
+
+@RunIf(min_gpus=1, deepspeed=True, special=True)
+def test_deepspeed_multigpu_stage_3_save_warning(tmpdir):
+    """
+    Test to ensure with Stage 3 and multiple GPUs that we recieve a warning that we're saving sharded checkpoints.
```
Comment (carmocca) [inline tests/plugins/test_deepspeed_plugin.py:647]: where are you checking this warning?

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 16 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:742]: They do in the comment weirdly enough, but not as an actual utility...

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 19]
```diff
@@ -657,43 +663,36 @@ def _create_default_config(
...
-            super().save_checkpoint(checkpoint, filepath)
-
-    def load_checkpoint_file(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
-        if self.save_full_weights or self.world_size == 1:
+        if self.zero_stage_3 and self._multi_device and self.is_global_zero:
+            # todo (sean): Add link to docs once docs are merged.
+            warning_cache.warn(
+                "When saving the DeepSpeed Stage 3 checkpoint, "
+                "each worker will save a shard of the checkpoint within a directory."
+                "If a single file is required after training, see <TODO> for instructions."
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/deepspeed.py:686]:
```suggestion
                "When saving the DeepSpeed Stage 3 checkpoint, "
                "each worker will save a shard of the checkpoint within a directory. "
                "If a single file is required after training, see <TODO> for instructions."
```

Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/deepspeed.py:686]: will this become a documentation entry?

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 20]
```diff
@@ -657,43 +663,36 @@ def _create_default_config(
...
-                    # State dict keys will include reference to wrapper LightningDeepSpeedModule
-                    # Delete `module` prefix before saving.
-                    state_dict = {k.partition("module.")[2]: state_dict[k] for k in state_dict.keys()}
-                    checkpoint["state_dict"] = state_dict
-                    return super().save_checkpoint(checkpoint, filepath)
-                return
-
-            # Use deepspeed's internal checkpointing function to handle partitioned weights across processes
-            # dump states as a checkpoint dictionary object
-            save_dir = self._filepath_to_dir(filepath)
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/deepspeed.py:688]: filepath_to_dir is not necessary anymore? deepspeed does that?

Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/deepspeed.py:688]: perhaps the function can be deleted then

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 21]
```diff
@@ -703,20 +702,77 @@ def load_checkpoint_file(self, checkpoint_path: Union[str, Path]) -> Dict[str, A
...
         )
+        if client_state is None:
+            raise MisconfigurationException(
+                "DeepSpeed was unable to load the checkpoint. Ensure you passed in a DeepSpeed compatible checkpoint "
+                "or a single checkpoint file with `Trainer(plugins=DeepSpeedPlugin(load_full_weights=True))`."
+            )
         return client_state
 
+    @property
+    def lightning_restore_optimizer_and_schedulers(self) -> bool:
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/deepspeed.py:716]: I'm not able to find where this is used

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 15 above)
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/deepspeed.py:738]:
```suggestion
        This is then automatically synced across processes.
        
        Args:
```

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 19 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:686]: Yep will be. I need to still make the documentation reference in a separate PR to refer here!

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 20 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:688]: It has been deleted I think?

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 21 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:716]: https://github.com/PyTorchLightning/pytorch-lightning/blob/7a1e97203e947d2d9b3eae8e7b54fdedb4c7681c/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L226

File: .azure-pipelines/gpu-tests.yml [hunk 22]
```diff
@@ -51,7 +51,7 @@ jobs:
     - bash: |
         python -c "fname = 'requirements/extra.txt' ; lines = [line for line in open(fname).readlines() if 'horovod' not in line] ; open(fname, 'w').writelines(lines)"
         pip install fairscale>=0.3.4
-        pip install "deepspeed>=0.4.0, !=0.4.4"  # FIXME: bug with 0.4.4
+        pip install "deepspeed>=0.4.3, !=0.4.4"  # FIXME: bug with 0.4.4
```
Comment (kaushikb11) [inline .azure-pipelines/gpu-tests.yml:54]: @SeanNaren Thomas did a similar fix in #8674?

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 9 above)
Comment (kaushikb11) [reply pytorch_lightning/callbacks/model_checkpoint.py:528]: Sean, is there a test for this?

File: .azure-pipelines/gpu-tests.yml (hunk 22 above)
Comment (SeanNaren) [reply .azure-pipelines/gpu-tests.yml:54]: Yeah I think we'll overwrite this PR's with Thomas's as we try to fix the issue! Whichever gets merged first :)

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 23]
```diff
@@ -657,43 +663,36 @@ def _create_default_config(
...
 
-    def _filepath_to_dir(self, filepath: str) -> str:
-        return os.path.dirname(filepath)
-
     @property
     def deepspeed_engine(self):
         return self.model
 
+    @property
+    def _multi_device(self) -> bool:
```
Comment (kaushikb11) [inline pytorch_lightning/plugins/training_type/deepspeed.py:671]: Should this be `multi_node` instead?

Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:671]: it checks both, so maybe it just should be `is_distributed`?

Comment (kaushikb11) [reply pytorch_lightning/plugins/training_type/deepspeed.py:671]: Yup, as we already have a property for that in `TrainingTypePlugin`

File: .azure-pipelines/gpu-tests.yml (hunk 22 above)
Comment (kaushikb11) [reply .azure-pipelines/gpu-tests.yml:54]: It's merged. I think the rebase would work?

File: tests/plugins/test_deepspeed_plugin.py [hunk 24]
```diff
@@ -419,6 +417,54 @@ def test_deepspeed_fp32_works(tmpdir):
...
+    plugin = trainer.training_type_plugin
+    assert isinstance(plugin, DeepSpeedPlugin)
+    assert not plugin.load_full_weights
+    with pytest.raises(MisconfigurationException, match="DeepSpeed was unable to load the checkpoint."):
+        trainer.test(model, ckpt_path=checkpoint_path)
+
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        plugins=[DeepSpeedPlugin(stage=3, load_full_weights=True)],
+        gpus=1,
```
Comment (kaushikb11) [inline tests/plugins/test_deepspeed_plugin.py:458]: The test mentions multi gpus, but single gpu being used?

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 23 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:671]: ahh didn't know that, hesitant to overwrite the function as it's used to select distributed kwargs and stuff I think. Maybe its best for now to keep it as is?

File: tests/plugins/test_deepspeed_plugin.py (hunk 24 above)
Comment (SeanNaren) [reply tests/plugins/test_deepspeed_plugin.py:458]: Great catch!

File: pytorch_lightning/plugins/training_type/deepspeed.py [hunk 25]
```diff
@@ -657,43 +663,36 @@ def _create_default_config(
...
+            warning_cache.warn(
+                "When saving the DeepSpeed Stage 3 checkpoint, "
+                "each worker will save a shard of the checkpoint within a directory. "
+                "If a single file is required after training, see <TODO> for instructions."
+            )
+        # Use deepspeed's internal checkpointing function to handle partitioned weights across processes
+        # dump states as a checkpoint dictionary object
+        _exclude_keys = ["state_dict", "optimizer_states", "lr_schedulers"]
+        checkpoint = {k: v for k, v in checkpoint.items() if k not in _exclude_keys}
+        self.deepspeed_engine.save_checkpoint(filepath, client_state=checkpoint)
```
Comment (kaushikb11) [inline pytorch_lightning/plugins/training_type/deepspeed.py:692]: As we are using DeepSpeed's internal checkpointing functionality, remote filesystems wouldn't be supported. Maybe we could mention it in a follow up PR?

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 23 above)
Comment (kaushikb11) [reply pytorch_lightning/plugins/training_type/deepspeed.py:671]: Sure, works. But would like to have a comment for future reference.

**Comment by kaushikb11** [comment]
Disabled auto-merge if you plan to push few more changes. LGTM!

File: .azure-pipelines/gpu-tests.yml (hunk 22 above)
Comment (SeanNaren) [reply .azure-pipelines/gpu-tests.yml:54]: oh i thought you were talking about https://github.com/PyTorchLightning/pytorch-lightning/pull/8655 haha

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/plugins/training_type/deepspeed.py (hunk 25 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/deepspeed.py:692]: Will add to docs, thank you!!

File: pytorch_lightning/callbacks/model_checkpoint.py (hunk 9 above)
Comment (SeanNaren) [reply pytorch_lightning/callbacks/model_checkpoint.py:528]: It's covered in the deepspeed save/load here: https://github.com/PyTorchLightning/pytorch-lightning/blob/fix/ds_saving_2/tests/plugins/test_deepspeed_plugin.py#L576
